### PR TITLE
rfc: introduce prelu primitive

### PR DESCRIPTION
**Pull request by piotrchmiel**
[Rendered document](https://github.com/piotrchmiel/oneDNN/tree/prelu_rfc/rfcs/20200106-prelu)

RFC concerns introduction of PRelu support in oneDNN. PReLU is specific variant of LeakyReLU, where alpha is learnable parameter. Full motivation is described inside document.

### CODE REVIEW

File: rfcs/20200106-prelu/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+Request comes from Caffe users: "PRelu is used in some POR topologies, such as
+VNET, SRGAN. It takes most of time if the original implementation, is not good
+optimized. For example VNET's train perf with latest Intel Caffe is 2.4s, PReLU
+(has optimized by openmp) spends 429ms (not including reorder time), 17% of total
+time. If we uses MKLDNN ReLU to make projection of the perf, it will come to
+~1.8s." Some other data scientist in discussion mentioned that he encouter PRelu
+issue when he tried to optimize the inference pipeline of a popular face-related
+application - PRelu tooks 47 % of time in his application.
+
+Currently in oneDnn we support ReLu activation function in eltwise operation:
```
Comment (kwiersch) [inline rfcs/20200106-prelu/README.md:15]: style: oneDNN

File: rfcs/20200106-prelu/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html. LeakyReLu can be
+achieved by specifying non zero alpha parameter. PReLu is specific variant of
+LeakyRelu, where alpha is learnable parameter. It cames with two flavours:
+- channel-wise PReLU - where number of trained alpha parameters is equal to
+channel number
+- channel-shared PReLu - where there is only one trainable alpha
```
Comment (kwiersch) [inline rfcs/20200106-prelu/README.md:21]: style, throughout: use a consistent capitalization scheme for "prelu" (probably PReLU is best)

File: rfcs/20200106-prelu/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t alpha_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Alpha gradient memory descriptor.
+    dnnl_memory_desc_t diff_alpha_desc;
+
+} dnnl_prelu_desc_t;
```
Comment (kwiersch) [inline rfcs/20200106-prelu/README.md:56]: Is PReLU the only parameterized eltwise operation of interest? Does it make sense to have a more generic parameterized eltwise primitive (or at least an API that allows for extension to new parameterized eltwise operations, so that a new primitive does not need to be added for the next one)?

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:56]: Currently I have no idea. I will check frameworks / contact people and go back to you with answer.

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:56]: I received feedback from PyTorch team: "For PyTorch side, I made some checks , but not find such eltwise ops which have learnable parameters."
I checked Caffe and I don't see also such primitives https://github.com/BVLC/caffe/tree/master/src/caffe/layers

File: rfcs/20200106-prelu/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+```
+Extend dnnl_alg_kind_t: dnnl_prelu_channel_wise, dnnl_prelu_channel_shared
+
+Add new defines:
+``` cpp
+#define DNNL_ARG_ALPHA number
+#define DNNL_ARG_DIFF_ALPHA number
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:63]: No need for the new names -- those should be just WEIGHTS

File: rfcs/20200106-prelu/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t alpha_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Alpha gradient memory descriptor.
+    dnnl_memory_desc_t diff_alpha_desc;
+
+} dnnl_prelu_desc_t;
+```
+Extend dnnl_alg_kind_t: dnnl_prelu_channel_wise, dnnl_prelu_channel_shared
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:58]: I think we should control the algorithm (single alpha, per-channel, or (TF-variant) the whole tensor) by reading the alpha descriptor with broadcast semantics:

``` cpp
md data = {{1, 2, 3, 4}, f32, format_tag};

// per tensor alpha (aka channel shared)
md alpha_per_tensor = {{1, 1, 1, 1}, f32, format_tag::any};

// per channel alpha (aka channel-wise)
md alpha_per_channel = {{1, 2, 1, 1}, f32, format_tag::any};

// per element alpha (whole tensor, TF/Keras-variant)
md alpha_per_element = {{1, 2, 3, 4}, f32, format_tag::any};
```

File: rfcs/20200106-prelu/README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+I would recommend option number A. In case of calculation gradients for alpha we
+ must pass memory descriptors for alpha and alpha_grad. Implementation
+ in existing eltwise kernel would require breaking the ABI. What is more now in
+ eltwise we pass float as alpha param. If we added DNNL_ARG_ALPHA required for
+ PReLU and alpha required for LeakyReLU remained it would be ambiguous for the user.
+
+## Open questions
+1. Tensorflow gives user greater flexibility than channel wise/shared,
+by shared_axes parameter. Should we support that as well ?
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:76]: IMO, API-wise yes, but the actual implementation can probably wait

File: rfcs/20200106-prelu/README.md (hunk 5 above)
Comment (emfomenk) [reply rfcs/20200106-prelu/README.md:58]: BTW, if we go this way, please specify the `format_tag::any` behavior for alpha: I think that would be the recommended option, so that we can use the format that would repeat the tensors format. Of course for single and per-channel cases this doesn't matter match, as the format will essentially be `abcd`.

File: rfcs/20200106-prelu/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+1. Option A - create new dedicated primitive PReLU:
+
+``` cpp
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_prelu.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training,
+    /// #dnnl_forward_inference, #dnnl_backward, and #dnnl_backward_data.
+    dnnl_prop_kind_t prop_kind;
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:42]: i don't think we need to support `backward_data`, only `backward = backward_data & backward_weights`.

File: rfcs/20200106-prelu/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    /// #dnnl_prelu_channel_shared.
+    dnnl_alg_kind_t alg_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t alpha_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Alpha gradient memory descriptor.
+    dnnl_memory_desc_t diff_alpha_desc;
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:54]: alpha --> weights

File: rfcs/20200106-prelu/README.md (hunk 3 above)
Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:56]: Answer from tensorflow team: "I checked activation functions in TF, PRelu is the only one that has learnable parameter"

File: rfcs/20200106-prelu/README.md (hunk 4 above)
Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md]: Fixed.

File: rfcs/20200106-prelu/README.md (hunk 5 above)
Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md]: I added paragraph and open question regarding broadcast, please review if our understanding of broadcast is the same.

File: rfcs/20200106-prelu/README.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
+
+Algorithm:
+- whole-tensor (tensorflow wariant) - weights_data = {{256, 7, 7, 128}, f32, nhwc};
+
+### Broadcast semantic
+
+Specifying format tag "any" for the weights tensor will enable broadcast feature.
```
Comment (vpirogov) [inline rfcs/20200106-prelu/README.md:68]: Why weights format is a condition for enabling broadcast?

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:68]: Maybe I misunderstood Everist comment above (https://github.com/oneapi-src/oneDNN/pull/737#discussion_r434018000) about specifying behaviour for `format_tag::any`. Generally broadcast feature can be enabled when it will match proposed "broadcastable" definition . If we go that way `format_tag::any` will only let us now that format of weights is exactly the same as source input. @emfomenk Maybe you meant something else about `format_tag::any` that I didn't catch ?

Comment (emfomenk) [reply rfcs/20200106-prelu/README.md:68]: I agree with @vpirogov that the broadcast semantic is orthogonal to the format tag (this statement is applicable for the entire library, and not only prelu primitive).

What I asked to clarify was what we would recommended users to use when creating a prelu primitive. Roughly, I would expect the following paragraph:
> The library will recommend user to allow prelu primitive to choose the appropriate weights memory format by passing `weights_md` with `format_tag::any`. This might be important for the full-tensor case, where  ideally the weights should repeat the memory format of the data tensor. However, given that the most typical case will be per-channel (C elements) or per-tensor (1 element) weights, the primitive should be optimized for the plain format that user has. Examples:
> - `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, tag::any}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, tag::any}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, c, 1, 1}, tag::abcd}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{1, 1, 1, 1}, tag::abcd}})` -- oK
> - `prelu_forward::desc(src_md, weights_md={{n, c, h, w}, tag::abcd}})` -- not recommended, since primitive might want to use `tab::aBcd16b`

As I said above, the broadcast semantic is orthogonal to the format. This is purely logical think and related to the logical dimensions only.

File: rfcs/20200106-prelu/README.md (hunk 5 above)
Comment (emfomenk) [reply rfcs/20200106-prelu/README.md]: Resolved in favor of https://github.com/oneapi-src/oneDNN/pull/737#discussion_r436106278

File: rfcs/20200106-prelu/README.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+} dnnl_prelu_desc_t;
+```
+For passing learnable parameters reuse existing defines:
+``` cpp
+#define DNNL_ARG_WEIGHTS number
+#define DNNL_ARG_DIFF_WEIGHTS number
+```
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:61]: (Conceptual issue detected!) Please note physical format (`nhwc` in this case) has nothing to do with the logical dimensions. For CNNs the order of logical dimensions is always `N, C, [[[D,] H,] W]`. So for this particular case the m_data should be declared as `{{256, 128, 7, 7}, f32, nhwc}`.

Some materials on that:
- https://oneapi-src.github.io/oneDNN/getting_started_cpp.html
- https://oneapi-src.github.io/oneDNN/dev_guide_understanding_memory_formats.html
- https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/20200115-new-memory-desc-operations/tensor-representation.md
    - As part of following RFC: https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200115-new-memory-desc-operations

File: rfcs/20200106-prelu/README.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+Algorithm:
+- whole-tensor (tensorflow wariant) - weights_data = {{256, 7, 7, 128}, f32, nhwc};
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:64]: Seems like an incomplete example. Please prettify. What is `m_data`, why only one bullet for algorithm, etc..

Comment (emfomenk) [reply rfcs/20200106-prelu/README.md:64]: Probably should go towards the end of option 1.

File: rfcs/20200106-prelu/README.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+            return false;
+    return true;
+}
+```
+After changed definition following case will be allowed:
+n = 256, h = 7, w = 7 c=128
+src_data = {{256, 7, 7, 128}, f32, nhwc};
+- weights_data = {{128, 1, 1, 128}, f32, any}
+
+Currently I don't see framework that support that, but technically it is doable.
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:137]: No. I don't think even NumPy supports such semantics. Given that there are other primitives that support broadcasting (matmul, binary), this option if considering, is a completely separate RFC. So I suggest drop this altogether from here.

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:137]: Ok.

File: rfcs/20200106-prelu/README.md (hunk 9 above)
Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md]: Ok, thank you for clarifying/ I added paragraph above. You used expression "appropriate weights". Do you see any case where primitive with weights description tagged as any should choose different data format than src_md format ? If no maybe I should specifying explicit that "any will let us now that format of weights is exactly the same as src md". I thought that maybe knowing format of weights can help somehow user with data initialisation or maybe in all cases user initializes the data with same || random values and it's format doesn't matter.

Comment (emfomenk) [reply rfcs/20200106-prelu/README.md]: > Do you see any case where primitive with weights description tagged as any should choose different data format than src_md format?

I don't, though the data types could be different.

> If no maybe I should specifying explicit that "any will let us now that format of weights is exactly the same as src md".

Let it be an implementation detail that we may want to document later in the "implementation detail" section of the documentation. No need to mention that in RFC though.

Comment (emfomenk) [reply rfcs/20200106-prelu/README.md]: Resolving the thread. Thanks!

**Comment by emfomenk** [comment]
@piotrchmiel, 
I updated the RFC with minor style and typo fixes. Hope you don't mind. Please review.

One more thing that came to my mind is specification for the backward pass. I added a new [open question](https://github.com/piotrchmiel/oneDNN/tree/prelu_rfc/rfcs/20200106-prelu#open-questions). Would you mind to check with FWKs if they are fine with the suggestion?

File: rfcs/20200106-prelu/README.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+   backward by data and by weights), or should it support those steps
+   separately?
+   - Separate pass might be required, if at some point the alpha is no more
+     treated as learnable parameter.
+   - It is not clear if cumulative backward pass brings any value, though not
+     providing it is strange, as other primitives (batch and layer
+     normalization) have it.
+
+   Current suggestion: support `backward = backward_data + backward_weights`,
+   and `backward_data`.
```
Comment (emfomenk) [inline rfcs/20200106-prelu/README.md:124]: One more thing that came to my mind is specification for the backward pass.
Would you mind to check with FWK folks if they are fine with the suggestion?

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:124]: Sure I will ask them.

Comment (piotrchmiel) [reply rfcs/20200106-prelu/README.md:124]: _For PyTorch side, it has just have one path(weight is always a learnable parameter): cumulative backward pass, see prelu_backward_cpu, so the suggestion is fine for me._
